craft_parts/filesystem_mounts.py

```diff
@@ -0,0 +1,163 @@
+
+    def __hash__(self) -> int:
+        return str.__hash__(self.mount)
+
+    def __eq__(self, other: object) -> bool:
+        if type(other) is type(self):
+            return self.mount == cast(FilesystemMountItem, other).mount
+
+        return False
+
+    @classmethod
+    def unmarshal(cls, data: dict[str, Any]) -> "FilesystemMountItem":
+        """Create and populate a new ``FilesystemMountItem`` object from dictionary data.
+
+        The unmarshal method validates entries in the input dictionary, populating
+        the corresponding fields in the data object.
+
+        :param data: The dictionary data to unmarshal.
+
+        :return: The newly created object.
+
+        :raise TypeError: If data is not a dictionary.
+        """
+        if not isinstance(data, dict):
+            raise TypeError("filesystem_mount item data is not a dictionary")
+
+        return FilesystemMountItem(**data)
+
+    def marshal(self) -> dict[str, Any]:
+        """Create a dictionary containing the filesystem_mount item data.
+
+        :return: The newly created dictionary.
+
+        """
+        return self.model_dump(by_alias=True)
+
+
+class FilesystemMount(RootModel):
+    """FilesystemMount defines the order in which devices should be mounted."""
+
+    root: Annotated[UniqueList[FilesystemMountItem], Field(min_length=1)]
+
+    @field_validator("root", mode="after")
+    @classmethod
+    def first_maps_to_slash(
+        cls, value: list[FilesystemMountItem]
+    ) -> list[FilesystemMountItem]:
+        """Make sure the first item in the list maps the '/' mount."""
+        if value[0].mount != "/":
+            raise ValueError("A filesystem first entry must map the '/' mount.")
+        return value
+
+    @classmethod
+    def unmarshal(cls, data: list[dict[str, Any]]) -> "FilesystemMount":
+        """Create and populate a new ``FilesystemMount`` object from list.
+
+        The unmarshal method validates entries in the input list, populating
+        the corresponding fields in the data object.
+
+        :param data: The list to unmarshal.
+
+        :return: The newly created object.
+
+        :raise TypeError: If data is not a list.
+        """
+        if not isinstance(data, list):
+            raise TypeError("filesystem_mount data is not a list")
+
+        return FilesystemMount(
+            root=[FilesystemMountItem.unmarshal(item) for item in data]
+        )
+
+    def marshal(self) -> list[dict[str, Any]]:
+        """Create a list containing the filesystem_mount data.
+
+        :return: The newly created list.
+
+        """
+        return cast(list[dict[str, Any]], self.model_dump(by_alias=True))
+
+
+FilesystemMounts = SingleEntryDict[Literal["default"], FilesystemMount]
+
+
+def validate_filesystem_mount(data: list[dict[str, Any]]) -> None:
+    """Validate a filesystem_mount.
+
+    :param data: The filesystem mount data to validate.
+    """
+    FilesystemMount.unmarshal(data)
```

Comment:
Should this have a try/except and raise `errors.FilesystemMountError`?  Otherwise I think this could raise a TypeError or pydantic error.

Since this and `validate_filesystem_mounts()` are designed to raise errors, a `:raises:` entry in the docstring would be useful.

Reply:
I improved the error handling. Let me know what you think about it.

Reply:
The handling for pydantic errors looks good but what would happen if `FilesystemMount.unmarshal` raises a `TypeError` on line 115?  Would this be caught on line 180?